src/application/common/logger.ts

```diff
@@ -0,0 +1,208 @@
+    getLogLevel(id: number): Promise<number>;
+    log(id: number, logLevel: number, message: string, params: any[]): Promise<void>;
+    child(obj: Object): Promise<number>;
+}
+
+export const ILoggerClient = Symbol("ILoggerClient");
+
+export interface ILogLevelChangedEvent {
+    oldLogLevel: number;
+    newLogLevel: number;
+}
+
+export interface ILoggerClient {
+    onLogLevelChanged(event: ILogLevelChangedEvent): void;
+}
+
+export const ILogger = Symbol("ILogger");
+
+export interface ILogger {
+    child(obj: Object): ILogger;
+    setLogLevel(logLevel: number): Promise<void>
+    getLogLevel(): Promise<number>;
+    trace(message: string, ...params: any[]): void;
+    debug(message: string, ...params: any[]): void;
+    info(message: string, ...params: any[]): void;
+    warn(message: string, ...params: any[]): void;
+    error(message: string, ...params: any[]): void;
+    fatal(message: string, ...params: any[]): void;
+}
+
+@injectable()
+export class Logger implements ILogger {
+
+    /* Log level for the loggers.  */
+    private _logLevel: Promise<number>[] = [];
+
+    /* Root logger has id 1.  */
+    private rootLoggerId = 1;
+
+    /* Default id is the root logger id.  */
+    private id: Promise<number> = Promise.resolve(this.rootLoggerId);
+
+    /**
+     * Build a new Logger.
+     *
+     * @param childOptions - The options to build the child with, see the
+     * bunyan documentation for more information.
+     */
+    constructor(
+        @inject(ILoggerServer) protected server: ILoggerServer,
+        @inject(LoggerWatcher) protected loggerWatcher: LoggerWatcher,
+        @unmanaged() childOptions?: Object) {
+
+        /* Creating a child logger.  */
+        if (childOptions !== undefined) {
+            this.id = server.child(childOptions);
+        }
+
+        /* Fetch the log level so it's cached in the frontend.  */
+        this.id.then(id => this._logLevel[id] = server.getLogLevel(id));
+
+        /* Update the root logger log level if it changes in the backend. */
+        loggerWatcher.onLogLevelChanged(event => {
+            this._logLevel[this.rootLoggerId] = Promise.resolve(event.newLogLevel);
+        });
+    }
+
+    /**
+     * Set the log level.
+     *
+     * @param loglevel - The loglevel to set. see Logger.LogLevel for
+     * possible options.
+     */
+    setLogLevel(logLevel: number): Promise<void> {
+        return new Promise<void>((resolve) => {
+            this.id.then(id => {
+                this._logLevel[id].then(oldLevel => {
+                    this.server.setLogLevel(id, logLevel).then(() => {
+                        this._logLevel[id] = Promise.resolve(logLevel);
+                        resolve();
+                    });
+                });
+            });
+        });
+    }
+
+    /**
+     * Get the log level.
+     *
+     * @returns a Promise to the log level.
+     */
+    getLogLevel(): Promise<number> {
+        return new Promise<number>(resolve => {
```

Comment:
Yeah part of it is just plain wrong I don't know how it got there like that... should be:
```typescript
   getLogLevel(): Promise<number> {
        return new Promise<number>(resolve => {
            this.id.then(id => resolve(this._logLevel[id]));
        });
    }
```

I should guard about a non existent id too..